src/main/java/org/tarantool/jdbc/SQLConnection.java

```diff
@@ -587,27 +520,40 @@ public void abort(Executor executor) throws SQLException {
     }
 
     @Override
-    public void setNetworkTimeout(Executor executor, int milliseconds) throws SQLException {
+    public int getNetworkTimeout() throws SQLException {
         checkNotClosed();
+        return (int) client.getOperationTimeout();
+    }
 
-        if (milliseconds < 0) {
-            throw new SQLException("Network timeout cannot be negative.");
-        }
+    protected SQLResultHolder execute(long timeout, String sql, Object... args) throws TimeoutException, SQLException {
+        checkNotClosed();
+        int networkTimeout = getNetworkTimeout();
+        return (timeout == 0 || (networkTimeout > 0 && networkTimeout < timeout))
+            ? executeWithNetworkTimeout(sql, args)
+            : executeWithStatementTimeout(timeout, sql, args);
+    }
 
+    private SQLResultHolder executeWithNetworkTimeout(String sql, Object... args) throws SQLException {
         try {
-            connection.setSocketTimeout(milliseconds);
-        } catch (SocketException e) {
-            throw new SQLException("Failed to set socket timeout: timeout=" + milliseconds, e);
+            return client.sqlRawOps().execute(sql, args);
+        } catch (Exception e) {
+            handleException(e);
+            throw new SQLException(formatError(sql, args), e);
         }
     }
 
-    @Override
-    public int getNetworkTimeout() throws SQLException {
-        checkNotClosed();
+    private SQLResultHolder executeWithStatementTimeout(long timeout, String sql, Object... args)
+        throws SQLException, TimeoutException {
         try {
-            return connection.getSocketTimeout();
-        } catch (SocketException e) {
-            throw new SQLException("Failed to retrieve socket timeout", e);
+            return client.sqlRawOps().execute(timeout, sql, args);
+        } catch (IllegalStateException e) {
+            if (e.getCause() instanceof TimeoutException) {
+                throw (TimeoutException) e.getCause();
+            }
+            throw new SQLException(formatError(sql, args), e);
```

Comment:
Maybe this logic would look better when in handleException(), what do you think? We'll need to introduce a class like StatementTimeoutException for that.

Reply:
Well, `handleException` was designed to perform some internal actions (i.e. close itself) depending on the error type. I think this method should not raise a new exception or propagate one. But here, we "override" the handling of TimeoutException (because `handleException` will close the connection when it consumes this exception type) and delegate handling to the call stack.
Inside the `handleException`, it's hard to distinguish whether it was aborted by network timeout or statement one because both are represented by TimeoutException. So, I decided that this method is responsible only for network timeout which impacts on the connection according to the spec.

Reply:
So there is no much need in StatementTimeoutException? Aside that, I think it worth to comment such places; the logic in not obvious at the first glance.

Reply:
No, I think StatementTimeoutException can be used to propagate it above. I commented on this piece of code to clarify the handling logic.

Reply:
Should not it be in SQLException hierarchy?

Reply:
Absolutely.